### PR TITLE
LPS-169414 KB: Add View child articles option to the dropdown

### PR DESCRIPTION
**Pull request by ambrinchaudhary**
Screenshot 2022-11-25 at 12 34 53

Based on what we have to show the link for child articles: 
```
<%
	int childKBArticlesCount = KBArticleServiceUtil.getKBArticlesCount(scopeGroupId, kbArticle.getResourcePrimKey(), WorkflowConstants.STATUS_ANY);
%>

	<c:if test="<%= childKBArticlesCount > 0 %>">
		<liferay-portlet:renderURL varImpl="childKBArticlesURL">
			<portlet:param name="mvcPath" value="/admin/view_kb_articles.jsp" />
			<portlet:param name="redirect" value="<%= currentURL %>" />
			<portlet:param name="resourcePrimKey" value="<%= String.valueOf(kbArticle.getResourcePrimKey()) %>" />
		        <portlet:param name="resourceClassNameId" value="<%= String.valueOf(kbArticle.getClassNameId()) %>" />
			<portlet:param name="parentResourceClassNameId" value="<%= String.valueOf(kbArticle.getClassNameId()) %>" />
			<portlet:param name="parentResourcePrimKey" value="<%= String.valueOf(kbArticle.getResourcePrimKey()) %>" />
			<portlet:param name="selectedItemId" value="<%= String.valueOf(kbArticle.getResourcePrimKey()) %>" />
		</liferay-portlet:renderURL>

		<span class="text-default">
			<aui:a href="<%= childKBArticlesURL.toString() %>">
				<liferay-ui:message arguments="<%= childKBArticlesCount %>" key="x-child-articles" />
			</aui:a>
		</span>
	</c:if>
```

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:sf

**Comment by liferay-lima** [comment]
ci:test:relevant

File: modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/util/KBDropdownItemsProvider.java [hunk 1]
```diff
@@ -910,6 +915,35 @@ private UnsafeConsumer<DropdownItem, Exception> _getRSSActionUnsafeConsumer(
 		};
 	}
 
+	private UnsafeConsumer<DropdownItem, Exception>
+		_getViewChildArticlesActionUnsafeConsumer(KBArticle kbArticle) {
```
Comment (marcogalluzzi) [inline modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/util/KBDropdownItemsProvider.java:919]: This method name is OK, but to make the other similar method name consistent I would rename `_getAddChildActionUnsafeConsumer` to `_getAddChildArticleActionUnsafeConsumer`

File: modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/util/KBDropdownItemsProvider.java [hunk 2]
```diff
@@ -1019,6 +1053,18 @@ private boolean _hasImportPermission(KBFolder kbFolder)
...
+			_themeDisplay.getScopeGroupId(), kbArticle.getResourcePrimKey(),
+			WorkflowConstants.STATUS_ANY);
+
+		if (childKBArticlesCount > 0) {
+			return true;
+		}
+
+		return false;
+	}
+
```
Comment (marcogalluzzi) [inline modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/util/KBDropdownItemsProvider.java:1067]: I would rename this method to `_hasChildArticles` and the variable to `childArticlesCount`.

**Review by marcogalluzzi — CHANGES_REQUESTED** [review]
I would allow the "View Child Articles" action only when navigating in KNOWLEDGE_BASE_ADMIN, it doesn't make sense having this option in KB Display widget. See attached video.

https://user-images.githubusercontent.com/19521293/204232229-7f69584d-1da0-4ff4-a99a-9af532b8c853.mp4

The change I would do is: create method `_hasViewChildArticlesPermission` and use it together with `_hasChildArticles` in `getKBArticleDropdownItems`.

**Review by marcogalluzzi — APPROVED** [review]

**Comment by ambrinchaudhary** [comment]
ci:test:sf

**Comment by ambrinchaudhary** [comment]
ci:test:relevant

**Comment by Hanlf** [comment]
ci:test:relevant

**Comment by Hanlf** [comment]
The above failures are not related to this PR.

**Comment by ambrinchaudhary** [comment]
@AliciaGarciaGarcia would you like to review this? otherwise I will forward it to Brian.

File: modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/util/KBDropdownItemsProvider.java [hunk 3]
```diff
@@ -973,6 +973,18 @@ private boolean _hasAddPermission() {
 		return false;
 	}
 
+	private boolean _hasChildArticles(KBArticle kbArticle) {
+		int childArticlesCount = KBArticleServiceUtil.getKBArticlesCount(
```
Comment (AliciaGarciaGarcia) [inline modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/util/KBDropdownItemsProvider.java:979]: my only question is about this change of name 

usually it's used _kbArticlesCount_ or something similar. but I do not know why you changed to _childArticlesCount_

It's a reason I miss?

same with `_hasKBChildArticles`

Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/util/KBDropdownItemsProvider.java:979]: To me it seems redundant to add the prefix `KB` when dealing with code in a module called _knowledge-base-web_ inside a class called _KBDropdownItemsProvider_ and having an input parameter called _kbArticle_ of type _KBArticle_. However, if guidelines say to use it I will do.

Comment (AliciaGarciaGarcia) [reply modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/util/KBDropdownItemsProvider.java:979]: It's better to change it.
And you do not have articles, you have kbArticles. 
An article can be of journal :S

Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/util/KBDropdownItemsProvider.java]: Renamed using standard guidelines, thanks for feedback

**Comment by marcogalluzzi** [comment]
ci:forward

**Comment by marcogalluzzi** [comment]
ci:reevaluate:1317401_6157

**Comment by ambrinchaudhary** [comment]
ci:reevaluate:1317401_6157

**Comment by ambrinchaudhary** [comment]
ci:forward

**Comment by Hanlf** [comment]
ci:forward

**Comment by Hanlf** [comment]
The above failures are not related to this PR, sent it to BChan here: https://github.com/brianchandotcom/liferay-portal/pull/127099

**Pull request closed without merging**